Tools/ML/ONNX/model.cxx

```diff
@@ -0,0 +1,213 @@
+  mEnv = std::make_shared<Ort::Env>(ORT_LOGGING_LEVEL_WARNING, "onnx-model");
+  mSession.reset(new Ort::Experimental::Session{*mEnv, path, sessionOptions});
+
+  mInputNames = mSession->GetInputNames();
+  mInputShapes = mSession->GetInputShapes();
+  mOutputNames = mSession->GetOutputNames();
+  mOutputShapes = mSession->GetOutputShapes();
+
+  LOG(info) << "Input Nodes:";
+  for (size_t i = 0; i < mInputNames.size(); i++) {
+    LOG(info) << "\t" << mInputNames[i] << " : " << printShape(mInputShapes[i]);
+  }
+
+  LOG(info) << "Output Nodes:";
+  for (size_t i = 0; i < mOutputNames.size(); i++) {
+    LOG(info) << "\t" << mOutputNames[i] << " : " << printShape(mOutputShapes[i]);
+  }
+
+  LOG(info) << "--- Model initialized! ---";
+
+  if (checkHyperloop()) {
+    sessionOptions.SetIntraOpNumThreads(activeThreads);
+  }
+}
+
+bool OnnxModel::fetchFromCCDB(std::string path_from, int64_t ccdbTimestamp, std::string path_to = "model.onnx")
+{
+
+  LOG(info) << "--- ONNX-ML model ---";
+
+  // e.g. path_from = "Analysis/PID/TPC/ML"
+
+  o2::ccdb::CcdbApi ccdbApi;
+  ccdbApi.init(ccdbUrl);
+
+  std::map<std::string, std::string> metadata;
+  bool retrieve_success = ccdbApi.retrieveBlob(path_from, ".", metadata, ccdbTimestamp, false, path_to);
+  std::map<std::string, std::string> headers = ccdbApi.retrieveHeaders(path_from, metadata, ccdbTimestamp);
+
+  if (headers.count("Valid-From") == 0) {
+    LOG(info) << "Valid-From not found in metadata";
+  } else {
+    valid_until = strtoul(headers["Valid-From"].c_str(), NULL, 0);
+  }
+  if (headers.count("Valid-Until") == 0) {
+    LOG(info) << "Valid-Until not found in metadata";
+  } else {
+    strtoul(headers["Valid-Until"].c_str(), NULL, 0);
+  }
+
+  if (checkHyperloop()) {
+    sessionOptions.SetIntraOpNumThreads(activeThreads);
+  }
+
+  mEnv = std::make_shared<Ort::Env>(ORT_LOGGING_LEVEL_WARNING, "onnx-model");
+  mSession.reset(new Ort::Experimental::Session{*mEnv, path_to, sessionOptions});
+
+  return retrieve_success;
+}
+
+bool OnnxModel::downloadToFile(std::string path_from, int64_t ccdbTimestamp, std::string path_to = "model.onnx")
+{
+
+  // e.g. path_from = "Analysis/PID/TPC/ML"
+
+  o2::ccdb::CcdbApi ccdbApi;
+  ccdbApi.init(ccdbUrl);
+
+  std::map<std::string, std::string> metadata;
+  bool retrieve_success = ccdbApi.retrieveBlob(path_from, ".", metadata, ccdbTimestamp, false, path_to);
+  std::map<std::string, std::string> headers = ccdbApi.retrieveHeaders(path_from, metadata, ccdbTimestamp);
+
+  if (headers.count("Valid-From") == 0) {
+    LOG(info) << "Valid-From not found in metadata";
+  } else {
+    LOG(info) << "Timestamp, valid from: " << headers["Valid-From"].c_str();
+  }
+  if (headers.count("Valid-Until") == 0) {
+    LOG(info) << "Valid-Until not found in metadata";
+  } else {
+    LOG(info) << "Timestamp, valid until: " << headers["Valid-Until"].c_str();
+  }
+
+  return retrieve_success;
+}
+
+float* OnnxModel::evalModel(std::vector<Ort::Value> input)
```

Comment:
Would it be more intuitive for the users to get float instead of a pointer?
Do you have multiple output values (a vector) in your network?

In the future, we can extend it to return any type, for now ok, float is fine.

Reply:
also, what is the use case for calling `evalModel` with an ONNX vector?

Reply:
> Would it be more intuitive for the users to get float instead of a pointer? Do you have multiple output values (a vector) in your network?

Indeed, I usually have multiple output values since I call the run function on a vector containing multiple entries (e.g. tracks). The float* returns a float array in this case. For a single return value, a float would also be fine, but for optimization (calling the ONNX Run()-function once instead of N-times) I considered returning a float.

> In the future, we can extend it to return any type, for now ok, float is fine.

That would be great! Maybe a template instance would be best here...

Reply:
> also, what is the use case for calling `evalModel` with an ONNX vector?

For now, it's only a feature if someone wants to make modifications to the tensor before passing it to the model. It's not needed now, but might be helpful in the future.

Reply:
Ok, I see your point.
Long time ago I could not get multiple outputs from ONNX. So, it's great it's possible!

Can you point me how do you then push multiple tracks into ONNX? You probably need to manually extract track columns and copy values for ONNX?

Reply:
Sure, you can find it e.g. here: https://github.com/AliceO2Group/O2Physics/blob/7234c430df8a1e90b893ef6da7b60edd586d3bd2/Common/TableProducer/PID/pidTPCFull.cxx#L278-L317
One has to be a bit careful then not to mess up the indexing, but otherwise it should work.

Reply:
Oh, so it is a flat array? And onnx itself can make it a 2D tensor or something like that, for the input?

Reply:
Sort of, yeah.
So ONNX (I think) checks the input dimension of the model and then passes as many values from the array into it as possible. So if your model has k inputs and you generate them for N tracks, then your array is k x N long. ONNX then produces a total N outputs for this input. So it automatically applies everything based on its input dimensions.

Reply:
I see, thanks! I see now that you yourself create the 2D tensor inside `evalModel()`, that makes sense now!
For sure, after the first PRs we will need to test this on more cases (what if `mInputShapes[0][1]==0`?), for now it is fine. I am getting a bit spammy, using this GitHub conversation to save any potential extension ideas for the future ;-)